programs/witness_node/main.cpp

```diff
@@ -157,27 +156,16 @@ int main(int argc, char** argv) {
    {
       elog("Exiting with error:\n${e}", ("e", unhandled_exception->to_detail_string()));
 
-      chain::block_database block_id_to_block;
-      block_id_to_block.open(data_dir / "database" / "block_num_to_block");
-      auto last_block = block_id_to_block.last();
-      fc::optional<chain::block_id_type> last_block_id = block_id_to_block.last_id();
       /*
          Possible to save unclean object database to file during replay (https://github.com/bitshares/bitshares-core/issues/946)
          During replay, undo_db is disabled, that said, when there is an exception thrown, 
          changes made to object database won't rewound, which means the in-memory object database would be unclean.
          When caught the exception, the node will dump object database from memory to disk. That means the on-disk object 
          database would be unclean. On next (and future) startup, if does not specify --replay-blockchain, 
          the on-disk object database would be considered clean and be loaded into memory directly, that's why we need to
-         pass do_flush = FALSE in event of exception while rebuilding object graph by replaying all blocks.
+         pass do_flush = FALSE in event of exception while undo_db is disabled.
       */
-      if
-      ( 
-         options.count("replay-blockchain") || 
-         (
-            last_block && last_block_id.valid() && *last_block_id >= node->chain_database()->head_block_id() &&
-            last_block->block_num() > node->chain_database()->head_block_num()
-         )
-      )
+      if( node->chain_database()->undo_db_disabled() )
          node->shutdown(false);
```

Comment:
I don't think it's good to deal with `database` in `main()`. IMHO a better approach would be telling `application` either "we got an error, please shutdown" or "we are fine, please shutdown", and let `application` decide what to do. Then, perhaps `application` doesn't know what to do either and will pass the info to `database` and let `database` decide, or perhaps it knows what to do and do it accordingly.

Reply:
Thanks !

Reply:
Perhaps I think the best way to put this logic inside application and not inside database, IMHO application must know if there was a exception during replay mode and that database must not be flushed. I think this logic in application makes more sense than if it would be in database.